### PR TITLE
Cover `AddSubmenu` with Unit tests

### PR DESCRIPTION
**Pull request by LevchenkoS**
Covered `AddSubMenuResponse` and
`AddSubMenuRequest` with Unit tests.

Related to: [APPLINK-25331](https://adc.luxoft.com/jira/browse/APPLINK-25331)

@AByzhynar, @okozlovlux, @Kozoriz, @LuxoftAKutsan, @vlantonov, @wolfylambova22, please review.

### CODE REVIEW

File: src/components/application_manager/src/commands/mobile/add_sub_menu_request.cc [hunk 1]
```diff
@@ -59,7 +59,7 @@ void AddSubMenuRequest::Run() {
 
   const int32_t menu_id =
       (*message_)[strings::msg_params][strings::menu_id].asInt();
-  if (app->FindSubMenu(menu_id)) {
+  if (!app->FindSubMenu(menu_id)) {
```
Comment (AByzhynar) [inline src/components/application_manager/src/commands/mobile/add_sub_menu_request.cc]: @LevchenkoS Shuould be done within separate task or as minimum different commit

File: src/components/application_manager/src/commands/mobile/add_sub_menu_request.cc [hunk 2]
```diff
@@ -60,7 +60,7 @@ void AddSubMenuRequest::Run() {
   const int32_t menu_id =
       (*message_)[strings::msg_params][strings::menu_id].asInt();
   if (app->FindSubMenu(menu_id)) {
-    SDL_ERROR("Menu with id " << menu_id << " is not found.");
+    SDL_ERROR("Menu with id " << menu_id << " is already exist.");
```
Comment (vlantonov) [inline src/components/application_manager/src/commands/mobile/add_sub_menu_request.cc]: @LevchenkoS I think that "already exists." or "is already existing." sounds better.

File: src/components/application_manager/test/commands/mobile/add_sub_menu_request_test.cc [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+  Event event(hmi_apis::FunctionID::INVALID_ENUM);
+  CommandPtr command(CreateCommand<AddSubMenuRequest>());
+
+  EXPECT_CALL(app_mngr_, ManageMobileCommand(_, _)).Times(0);
+
+  command->on_event(event);
+}
+
+TEST_F(AddSubMenuRequestTest, OnEvent_SUCCESS) {
+  mobile_apis::Result::eType kResult = mobile_apis::Result::SUCCESS;
```
Comment (vlantonov) [inline src/components/application_manager/test/commands/mobile/add_sub_menu_request_test.cc]: @LevchenkoS Considering the name it seems that the `const` qualifier is missing.

File: src/components/application_manager/src/commands/mobile/add_sub_menu_request.cc (hunk 1 above)
Comment (LevchenkoS) [inline src/components/application_manager/src/commands/mobile/add_sub_menu_request.cc]: @AByzhynar, removed in 711c3da because of initial misunderstanding of the meaning of this expression.

File: src/components/application_manager/test/commands/mobile/add_sub_menu_request_test.cc (hunk 3 above)
Comment (LevchenkoS) [inline src/components/application_manager/test/commands/mobile/add_sub_menu_request_test.cc]: @vlantonov, fixed in 711c3da.

File: src/components/application_manager/src/commands/mobile/add_sub_menu_request.cc [hunk 4]
```diff
@@ -60,7 +60,7 @@ void AddSubMenuRequest::Run() {
   const int32_t menu_id =
       (*message_)[strings::msg_params][strings::menu_id].asInt();
   if (app->FindSubMenu(menu_id)) {
-    SDL_ERROR("Menu with id " << menu_id << " is not found.");
+    SDL_ERROR("Menu with id " << menu_id << " is already exists.");
```
Comment (vlantonov) [inline src/components/application_manager/src/commands/mobile/add_sub_menu_request.cc]: @LevchenkoS This was not changed since my previous comment although you claim it is fixed. I meant that " is already exists." is not correct. Better suggestions are "already exists." or "is already existing."

Comment (LevchenkoS) [inline src/components/application_manager/src/commands/mobile/add_sub_menu_request.cc]: @vlantonov, sorry. I misunderstood your comment. Fixed in 62dadbd.

**Comment by vlantonov** [comment]
@LevchenkoS Reviewed.

File: src/components/application_manager/test/commands/mobile/add_sub_menu_request_test.cc [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+
+using ::testing::_;
+using ::testing::Return;
+using ::testing::ReturnRef;
+namespace am = ::application_manager;
+using am::commands::AddSubMenuRequest;
+using am::commands::MessageSharedPtr;
+using am::event_engine::Event;
+
+typedef SharedPtr<AddSubMenuRequest> CommandPtr;
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/test/commands/mobile/add_sub_menu_request_test.cc]: @LevchenkoS  I would prefer name this AddSubMenuPtr. 
CommandPtr is a bit confusing

File: src/components/application_manager/test/commands/mobile/add_sub_menu_request_test.cc [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+  const mobile_apis::Result::eType kResult = mobile_apis::Result::SUCCESS;
+  Event event(hmi_apis::FunctionID::UI_AddSubMenu);
+  MessageSharedPtr event_msg(CreateMessage(smart_objects::SmartType_Map));
+  (*event_msg)[am::strings::params][am::hmi_response::code] = kResult;
+  (*event_msg)[am::strings::msg_params] = 0;
+
+  event.set_smart_object(*event_msg);
+
+  MockAppPtr app(CreateMockApp());
+  ON_CALL(app_mngr_, application(_)).WillByDefault(Return(app));
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/test/commands/mobile/add_sub_menu_request_test.cc]: @LevchenkoS  do you rely don't care what application id will be passed?

File: src/components/application_manager/test/commands/mobile/add_sub_menu_request_test.cc [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+  (*command_msg)[am::strings::msg_params][am::strings::menu_name] =
+      "valid_sub_menu_name";
+  (*command_msg)[am::strings::msg_params][am::strings::position] =
+      "test_position";
+
+  CommandPtr command(CreateCommand<AddSubMenuRequest>(command_msg));
+
+  MockAppPtr app(CreateMockApp());
+  ON_CALL(app_mngr_, application(_)).WillByDefault(Return(app));
+  ON_CALL(*app, FindSubMenu(_))
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/test/commands/mobile/add_sub_menu_request_test.cc]: @LevchenkoS I think tests should test that AddSubmenu is trying ti find correct sub menu in correct application

File: src/components/application_manager/test/commands/mobile/add_sub_menu_request_test.cc (hunk 6 above)
Comment (LevchenkoS) [inline src/components/application_manager/test/commands/mobile/add_sub_menu_request_test.cc]: @LuxoftAKutsan, in this case - yes. Because command will take application id  form it's own message.

**Comment by wolfylambova22** [comment]
Reviewed.

**Comment by okozlovlux** [comment]
@LevchenkoS reviewed

File: src/components/application_manager/test/commands/mobile/add_sub_menu_request_test.cc (hunk 7 above)
Comment (LevchenkoS) [inline src/components/application_manager/test/commands/mobile/add_sub_menu_request_test.cc]: @LuxoftAKutsan, fixed in 34fd2f1.

File: src/components/application_manager/test/commands/mobile/add_sub_menu_request_test.cc (hunk 5 above)
Comment (LevchenkoS) [inline src/components/application_manager/test/commands/mobile/add_sub_menu_request_test.cc]: @LuxoftAKutsan, fixed in 34fd2f1.

File: src/components/application_manager/test/commands/mobile/add_sub_menu_request_test.cc [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+
+  MockAppPtr app(CreateMockApp());
+  EXPECT_CALL(app_mngr_, application(kConnectionKey)).WillOnce(Return(app));
+  EXPECT_CALL(*app, FindSubMenu(kMenuId))
+      .WillOnce(Return(static_cast<SmartObject*>(NULL)));
+
+  MessageSharedPtr result_msg(CatchHMICommandResult(CallRun(*command)));
+  EXPECT_EQ(hmi_apis::FunctionID::UI_AddSubMenu,
+            static_cast<hmi_apis::FunctionID::eType>(
+                (*result_msg)[am::strings::params][am::strings::function_id]
```
Comment (Kozoriz) [inline src/components/application_manager/test/commands/mobile/add_sub_menu_request_test.cc]: @LevchenkoS , can you move this to variable initialization and to EXPECT put only var name?

Comment (Kozoriz) [inline src/components/application_manager/test/commands/mobile/add_sub_menu_request_test.cc]: @LevchenkoS, please correct in all test cases

File: src/components/application_manager/test/commands/mobile/add_sub_menu_request_test.cc (hunk 6 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/test/commands/mobile/add_sub_menu_request_test.cc]: @LevchenkoS  Where does it gets it own message?

Comment (LevchenkoS) [inline src/components/application_manager/test/commands/mobile/add_sub_menu_request_test.cc]: @LuxoftAKutsan, all commands gets their own message in constructor.

Comment (LuxoftAKutsan) [inline src/components/application_manager/test/commands/mobile/add_sub_menu_request_test.cc]: @LevchenkoS  you should check that command ask same application_id as in smart_object

File: src/components/application_manager/test/commands/mobile/add_sub_menu_request_test.cc (hunk 8 above)
Comment (LevchenkoS) [inline src/components/application_manager/test/commands/mobile/add_sub_menu_request_test.cc]: @Kozoriz, fixed in b128e93.

File: src/components/application_manager/test/commands/mobile/add_sub_menu_request_test.cc (hunk 6 above)
Comment (LevchenkoS) [inline src/components/application_manager/test/commands/mobile/add_sub_menu_request_test.cc]: @LuxoftAKutsan, fixed in b128e93.

**Comment by LuxoftAKutsan** [comment]
Reviewed

**Comment by Kozoriz** [comment]
Reviewed

File: src/components/application_manager/test/commands/mobile/add_sub_menu_request_test.cc [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+  MockAppPtr app(CreateMockApp());
+  EXPECT_CALL(app_mngr_, application(kConnectionKey)).WillOnce(Return(app));
+
+  MessageSharedPtr result_msg(
+      CatchMobileCommandResult(CallOnEvent(*command, event)));
+  const mobile_apis::Result::eType kReceivedResult =
+      static_cast<mobile_apis::Result::eType>(
+          (*result_msg)[am::strings::msg_params][am::strings::result_code]
+              .asInt());
+  EXPECT_EQ(kResult, kReceivedResult);
```
Comment (AByzhynar) [inline src/components/application_manager/test/commands/mobile/add_sub_menu_request_test.cc]: @LevchenkoS Remove `kResult` and use `mobile_apis::Result::XXXXXX` as you did in other test cases.

Comment (LevchenkoS) [inline src/components/application_manager/test/commands/mobile/add_sub_menu_request_test.cc]: @AByzhynar, fixed in 10df63f.

**Comment by AByzhynar** [comment]
Reviewed
